PreLoad.cmake

```diff
@@ -96,5 +96,16 @@ if (OS MATCHES "Linux"
     else ()
         message (FATAL_ERROR "Unsupported architecture: ${ARCH}")
     endif ()
+elseif (OS MATCHES "Darwin"
+    AND NOT DEFINED CMAKE_TOOLCHAIN_FILE
+    AND NOT DISABLE_HERMETIC_BUILD
+    AND ("$ENV{CC}" MATCHES ".*clang.*" OR CMAKE_C_COMPILER MATCHES ".*clang.*"))
 
+    if (ARCH MATCHES "amd64|x86_64")
+        set (CMAKE_TOOLCHAIN_FILE "cmake/darwin/toolchain-x86_64.cmake" CACHE INTERNAL "")
```

Comment:
Hm, toolchain usually used for cross compiling, but here you are checking for `OS=Darwin`, so it will match when it is executed under macos, but does macos has `x86_64-apple-darwin` compiler (this is what toolchain requests)?

Reply:
Right now the predominant architecture is ARM but all pre-2020 OSes are on x86 (see https://en.wikipedia.org/wiki/Mac_transition_to_Apple_silicon)

Reply:
Not sure that I got this, I mean that this toolchains for cross compiling from linux, as far as I know.

Reply:
Maybe it's something that I'm doing wrong, let me expand on the issue.

The issue's setup steps are checking out the current head, initialising submodules and doing ``mkdir build && cd build && cmake ..``.

The CMake configuration errors out as seen below as it appears ``CMAKE_TOOLCHAIN_FILE`` is unset
```
Homebrew clang version 17.0.6
Target: arm64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /opt/homebrew/opt/llvm/bin
-- Using linker: /usr/bin/ld
-- Using archiver: /opt/homebrew/opt/llvm/bin/llvm-ar
-- Using ranlib: /opt/homebrew/opt/llvm/bin/llvm-ranlib
-- Using install-name-tool: /opt/homebrew/opt/llvm/bin/llvm-install-name-tool
-- Using objcopy: /opt/homebrew/opt/llvm/bin/llvm-objcopy
-- Using strip: /opt/homebrew/opt/llvm/bin/llvm-strip
-- Using ccache: /opt/homebrew/bin/ccache (version 4.9)
-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-145)")
-- Git HEAD commit hash: 6858d2f4ca29b68a484e4f0845deb084ff6e9b0a
On branch master
Your branch is up to date with 'origin/master'.

...

-- Adding contrib module corrosion (configuring with corrosion-cmake)
-- Rust Target: aarch64-apple-darwin
-- Found Rust: /opt/homebrew/bin/rustc (found version "1.75.0")
-- Checking Rust toolchain for current target
CMake Error at contrib/corrosion-cmake/CMakeLists.txt:38 (message):
  Unsupported rust target
```

When setting it in ``Preload.cmake`` using the patch, or running ``cmake -DCMAKE_TOOLCHAIN_FILE=aarch64-apple-darwin ..``, the CMake configuration step instead completes with
```
-- Adding contrib module corrosion (configuring with corrosion-cmake)
-- Rust Target: aarch64-apple-darwin
-- Found Rust: /opt/homebrew/bin/rustc (found version "1.75.0") 
-- Checking Rust toolchain for current target
-- Switched Rust target to aarch64-apple-darwin
-- Using Corrosion as a subdirectory
``` 

What is the recommended way to configure the build system? Agree that technically we don't want to cross-compile. But it seems that's the mechanism currently used by ``corrosion-cmake``.

Reply:
Ok, understood, corrosion always assume cross compiling, can you test the following patch?

```patch
diff --git a/contrib/corrosion-cmake/CMakeLists.txt b/contrib/corrosion-cmake/CMakeLists.txt
index 9b98ed6efb3..697bcd149fc 100644
--- a/contrib/corrosion-cmake/CMakeLists.txt
+++ b/contrib/corrosion-cmake/CMakeLists.txt
@@ -16,26 +16,28 @@ message(STATUS "Checking Rust toolchain for current target")

 # See https://doc.rust-lang.org/nightly/rustc/platform-support.html

-if(CMAKE_TOOLCHAIN_FILE MATCHES "ppc64le")
-    set(Rust_CARGO_TARGET "powerpc64le-unknown-linux-gnu")
-elseif((CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-x86_64") AND (CMAKE_TOOLCHAIN_FILE MATCHES "musl"))
-    set(Rust_CARGO_TARGET "x86_64-unknown-linux-musl")
-elseif(CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-x86_64")
-    set(Rust_CARGO_TARGET "x86_64-unknown-linux-gnu")
-elseif((CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-aarch64") AND (CMAKE_TOOLCHAIN_FILE MATCHES "musl"))
-    set(Rust_CARGO_TARGET "aarch64-unknown-linux-musl")
-elseif(CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-aarch64")
-    set(Rust_CARGO_TARGET "aarch64-unknown-linux-gnu")
-elseif((CMAKE_TOOLCHAIN_FILE MATCHES "darwin") AND (CMAKE_TOOLCHAIN_FILE MATCHES "x86_64"))
-    set(Rust_CARGO_TARGET "x86_64-apple-darwin")
-elseif((CMAKE_TOOLCHAIN_FILE MATCHES "darwin") AND (CMAKE_TOOLCHAIN_FILE MATCHES "aarch64"))
-    set(Rust_CARGO_TARGET "aarch64-apple-darwin")
-elseif((CMAKE_TOOLCHAIN_FILE MATCHES "freebsd") AND (CMAKE_TOOLCHAIN_FILE MATCHES "x86_64"))
-    set(Rust_CARGO_TARGET "x86_64-unknown-freebsd")
-elseif(CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-riscv64")
-    set(Rust_CARGO_TARGET "riscv64gc-unknown-linux-gnu")
-else()
-    message(FATAL_ERROR "Unsupported rust target")
+if(DEFINED CMAKE_TOOLCHAIN_FILE)
+    if(CMAKE_TOOLCHAIN_FILE MATCHES "ppc64le")
+        set(Rust_CARGO_TARGET "powerpc64le-unknown-linux-gnu")
+    elseif((CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-x86_64") AND (CMAKE_TOOLCHAIN_FILE MATCHES "musl"))
+        set(Rust_CARGO_TARGET "x86_64-unknown-linux-musl")
+    elseif(CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-x86_64")
+        set(Rust_CARGO_TARGET "x86_64-unknown-linux-gnu")
+    elseif((CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-aarch64") AND (CMAKE_TOOLCHAIN_FILE MATCHES "musl"))
+        set(Rust_CARGO_TARGET "aarch64-unknown-linux-musl")
+    elseif(CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-aarch64")
+        set(Rust_CARGO_TARGET "aarch64-unknown-linux-gnu")
+    elseif((CMAKE_TOOLCHAIN_FILE MATCHES "darwin") AND (CMAKE_TOOLCHAIN_FILE MATCHES "x86_64"))
+        set(Rust_CARGO_TARGET "x86_64-apple-darwin")
+    elseif((CMAKE_TOOLCHAIN_FILE MATCHES "darwin") AND (CMAKE_TOOLCHAIN_FILE MATCHES "aarch64"))
+        set(Rust_CARGO_TARGET "aarch64-apple-darwin")
+    elseif((CMAKE_TOOLCHAIN_FILE MATCHES "freebsd") AND (CMAKE_TOOLCHAIN_FILE MATCHES "x86_64"))
+        set(Rust_CARGO_TARGET "x86_64-unknown-freebsd")
+    elseif(CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-riscv64")
+        set(Rust_CARGO_TARGET "riscv64gc-unknown-linux-gnu")
+    else()
+        message(FATAL_ERROR "Unsupported rust target")
+    endif()
 endif()

 message(STATUS "Switched Rust target to ${Rust_CARGO_TARGET}")
```